### PR TITLE
Add S3 Storage Class Analytics Configuration

### LINKED ISSUE #5: Support Analytics Configurations for S3 Buckets
(Migrating from hashicorp/terraform#13246)

AWS has recently announced some support for additional analytics and metrics for S3 buckets: https://aws.amazon.com/blogs/aws/s3-storage-management-update-analytics-object-tagging-inventory-and-metrics/

In order to support the above, the implementation in the AWS API is with methods like `GetBucketAnalyticsConfiguration` and `PutBucketAnalyticsConfiguration` which were added in go-aws-sdk 1.5.11.

### New Resource

- aws_s3_bucket_analytics

### Expected Behavior

Terraform configuration should support enabling and managing AWS S3 analytics configurations.

### References

* http://docs.aws.amazon.com/AmazonS3/latest/dev/analytics-storage-class.html
* https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.GetBucketAnalyticsConfiguration
* https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.PutBucketAnalyticsConfiguration
* https://github.com/aws/aws-sdk-go/commit/b4916cc32444fcca7ff773d1279b9c556c9cca10

### PR DESCRIPTION
**Pull request by gdavison**
Adds the resource `aws_s3_bucket_analytics_configuration` to enable [Storage Class Analysis](https://docs.aws.amazon.com/AmazonS3/latest/dev/analytics-storage-class.html) for S3 Buckets.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #5 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
**New Resource:** `aws_s3_bucket_analytics_configuration`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSS3BucketAnalyticsConfiguration_'

--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_Empty (3.48s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithStorageClassAnalysis_Empty (3.50s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_basic (28.49s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithStorageClassAnalysis_Full (29.08s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithStorageClassAnalysis_Default (29.08s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_removed (43.54s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_Remove (46.12s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_Prefix (46.21s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_PrefixAndTags (46.22s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_MultipleTags (46.32s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_SingleTag (46.39s)
--- PASS: TestAccAWSS3BucketAnalyticsConfiguration_updateBasic (67.60s)
```

### CODE REVIEW

File: aws/resource_aws_s3_bucket_analytics_configuration.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+					Schema: map[string]*schema.Schema{
+						"prefix": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							AtLeastOneOf: filterAtLeastOneOfKeys,
+						},
+						"tags": schemaWithAtLeastOneOf(
+							tagsSchema(),
+							filterAtLeastOneOfKeys,
+						),
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration.go]: Are there expectations to use `schemaWithAtLeastOneOf()` elsewhere? Seems like it is not adding more than a direct `AtLeastOneOf: <value>` to the schema. We generally prefer to explicitly declare the schema, to remove indirection for code readers and allow linters to check usage of the composite literals. 

```go
						"prefix": {
							Type:         schema.TypeString,
							Optional:     true,
							AtLeastOneOf: []string{"filter.0.prefix", "filter.0.tags"},
						},
						"tags": {
							Type:         schema.TypeMap,
							Optional:     true,
							Elem:         &schema.Schema{Type: schema.TypeString},
							AtLeastOneOf: []string{"filter.0.prefix", "filter.0.tags"},
						},
```

File: aws/resource_aws_s3_bucket_analytics_configuration.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+	log.Printf("[DEBUG] Reading S3 bucket analytics configuration: %s", input)
+	output, err := conn.GetBucketAnalyticsConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			log.Printf("[WARN] %s S3 bucket analytics configuration not found, removing from state.", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration.go]: Nit: Its generally a good idea to return error context about the error being returned for easier troubleshooting by operators and maintainers:

```suggestion
		return fmt.Errorf("error getting S3 Bucket Analytics Configuration (%s): %w", d.Id(), err)
```

File: aws/resource_aws_s3_bucket_analytics_configuration.go [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+	}
+
+	if err := d.Set("filter", flattenS3AnalyticsFilter(output.AnalyticsConfiguration.Filter)); err != nil {
+		return err
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration.go]: Nit: Same here and below

```suggestion
		return fmt.Errorf("error setting filter: %s", err)
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 4]
```diff
@@ -0,0 +1,1225 @@
...
+	})
+}
+
+func TestAccAWSS3BucketAnalyticsConfiguration_WithFilter_Empty(t *testing.T) {
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 5]
```diff
@@ -0,0 +1,1225 @@
...
+	resourceName := "aws_s3_bucket_analytics_configuration.test"
+
+	rName := fmt.Sprintf("tf-acc-test-%d", rInt)
+	prefix := fmt.Sprintf("prefix-%d/", rInt)
+	prefixUpdate := fmt.Sprintf("prefix-update-%d/", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 6]
```diff
@@ -0,0 +1,1225 @@
...
+	resourceName := "aws_s3_bucket_analytics_configuration.test"
+
+	rName := fmt.Sprintf("tf-acc-test-%d", rInt)
+	tag1 := fmt.Sprintf("tag-%d", rInt)
+	tag1Update := fmt.Sprintf("tag-update-%d", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 7]
```diff
@@ -0,0 +1,1225 @@
...
+	rName := fmt.Sprintf("tf-acc-test-%d", rInt)
+	tag1 := fmt.Sprintf("tag1-%d", rInt)
+	tag1Update := fmt.Sprintf("tag1-update-%d", rInt)
+	tag2 := fmt.Sprintf("tag2-%d", rInt)
+	tag2Update := fmt.Sprintf("tag2-update-%d", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 8]
```diff
@@ -0,0 +1,1225 @@
...
+	prefixUpdate := fmt.Sprintf("prefix-update-%d/", rInt)
+	tag1 := fmt.Sprintf("tag1-%d", rInt)
+	tag1Update := fmt.Sprintf("tag1-update-%d", rInt)
+	tag2 := fmt.Sprintf("tag2-%d", rInt)
+	tag2Update := fmt.Sprintf("tag2-update-%d", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 9]
```diff
@@ -0,0 +1,1225 @@
...
+	rInt := acctest.RandInt()
+	resourceName := "aws_s3_bucket_analytics_configuration.test"
+
+	rName := fmt.Sprintf("tf-acc-test-%d", rInt)
+	prefix := fmt.Sprintf("prefix-%d/", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 10]
```diff
@@ -0,0 +1,1225 @@
...
+	})
+}
+
+func TestAccAWSS3BucketAnalyticsConfiguration_WithStorageClassAnalysis_Empty(t *testing.T) {
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 11]
```diff
@@ -0,0 +1,1225 @@
...
+func TestAccAWSS3BucketAnalyticsConfiguration_WithStorageClassAnalysis_Default(t *testing.T) {
+	var ac s3.AnalyticsConfiguration
+	resourceName := "aws_s3_bucket_analytics_configuration.test"
+
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 12]
```diff
@@ -0,0 +1,1225 @@
...
+	resourceName := "aws_s3_bucket_analytics_configuration.test"
+
+	rInt := acctest.RandInt()
+	rName := fmt.Sprintf("tf-acc-test-%d", rInt)
+	prefix := fmt.Sprintf("prefix-%d/", rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketMetricDestroy,
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]:
```suggestion
		CheckDestroy: testAccCheckAWSS3BucketAnalyticsConfigurationDestroy,
```

File: aws/resource_aws_s3_bucket_analytics_configuration_test.go [hunk 13]
```diff
@@ -0,0 +1,1225 @@
...
+}
+
+resource "aws_s3_bucket" "test" {
+  bucket = "%s"
+}
+
+resource "aws_s3_bucket" "destination" {
+  bucket = "%s-destination"
+}
+`, name, prefix, bucket, bucket)
```
Comment (bflad) [inline aws/resource_aws_s3_bucket_analytics_configuration_test.go]: Nit: To save repeating fmt arguments, you can use indexed notation in the template, e.g. `%[3]s`

File: website/docs/r/s3_bucket_analysis_configuration.html.markdown [hunk 14]
```diff
@@ -0,0 +1,104 @@
+---
```
Comment (bflad) [inline website/docs/r/s3_bucket_analysis_configuration.html.markdown]: Documentation page is missing sidebar link in `website/aws.erb` (required until documentation is moved off terraform.io in the coming months)

Comment (gdavison) [reply website/docs/r/s3_bucket_analysis_configuration.html.markdown]: 🤦‍♂ Yeah, I've asked for that change a few times…

File: aws/resource_aws_s3_bucket_analytics_configuration.go (hunk 1 above)
Comment (gdavison) [reply aws/resource_aws_s3_bucket_analytics_configuration.go]: I was thinking in this case of leveraging the existing pre-defined tags schema. Even though it's not complicated, it's consistent across all resources.
Similar to how we have `tagsSchema()`, `tagsSchemaComputed()`, and `tagsSchemaForceNew()`, but making this a wrapper that could extend any existing schema with an `AtLeastOneOf`.

File: website/aws.erb [hunk 15]
```diff
@@ -2593,6 +2593,9 @@
                                 <li>
                                     <a href="/docs/providers/aws/r/s3_bucket.html">aws_s3_bucket</a>
                                 </li>
+                                <li>
+                                    <a href="/docs/providers/aws/r/s3_bucket_analysis_configuration.html">aws_s3_bucket_inventory</a>
```
Comment (bflad) [inline website/aws.erb]: 😎 

```suggestion
                                    <a href="/docs/providers/aws/r/s3_bucket_analysis_configuration.html">aws_s3_bucket_analysis_configuration</a>
```

**Pull request merged**